core/src/main/java/org/mule/runtime/core/DefaultEventContext.java

```diff
@@ -200,6 +237,29 @@ private DefaultEventContext(FlowConstruct flow, ComponentLocation location, Stri
     this.correlationId = correlationId;
   }
 
+  /**
+   * Builds a new execution context with the given parameters.
+   *
+   * @param id the unique id for this event context.
+   * @param location the location of the component that received the first message for this context.
+   * @param correlationId the correlation id that was set by the {@link MessageSource} for the first {@link Event} of this
+   *        context, if available.
+   * @param externalCompletionPublisher void publisher that completes when source completes enabling completion of
+   *        {@link EventContext} to depend on completion of source.
+   */
+  private DefaultEventContext(String id, ComponentLocation location, String correlationId,
+                              Publisher<Void> externalCompletionPublisher) {
+    super(externalCompletionPublisher);
+    this.id = id;
+    this.serverId = "-";
+    this.flowName = location.getParts().get(0).getPartPath();
```

Comment:
Refactor repeated use of this type of thing into static helpers in ComponentLocation.